### ISSUE TITLE
Possible bug in BaseNDArrayFactory.rand() when running with CUDA back-end

### ISSUE THREAD

**Issue by phvu** [deeplearning4j/nd4j#36]
The following program, when running with CUDA back-end, throws an error related to `cublasFreeNative()`. Both the program and error log are attached.
Configuration: Java 1.8, CUDA 6.5 on MacOS 10.10. CUDA Driver version 6.5.46, GPU Driver version 310.41
I tried the same program on Java 1.7, CUDA 6.5 on Amazon GPU instance (Grid K520) and it gives similar error.

Program
----
```
package org.deeplearning4j;

import org.deeplearning4j.datasets.iterator.DataSetIterator;
import org.deeplearning4j.datasets.iterator.MultipleEpochsIterator;
import org.deeplearning4j.datasets.iterator.impl.MnistDataSetIterator;
import org.deeplearning4j.eval.Evaluation;
import org.deeplearning4j.models.featuredetectors.rbm.RBM;
import org.deeplearning4j.nn.api.OptimizationAlgorithm;
import org.deeplearning4j.nn.conf.MultiLayerConfiguration;
import org.deeplearning4j.nn.conf.NeuralNetConfiguration;
import org.deeplearning4j.nn.conf.override.ClassifierOverride;
import org.deeplearning4j.nn.layers.factory.LayerFactories;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.deeplearning4j.nn.weights.WeightInit;
import org.deeplearning4j.optimize.listeners.ScoreIterationListener;
import org.nd4j.linalg.api.buffer.DataBuffer;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.rng.DefaultRandom;
import org.nd4j.linalg.api.rng.Random;
import org.nd4j.linalg.dataset.DataSet;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.lossfunctions.LossFunctions;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;

/**
 * Created by vupham on 3/10/15.
 */
public class DBNExample {
    private static Logger log = LoggerFactory.getLogger(DBNExample.class);


    public static void main(String[] args) throws Exception {
        long randomSeed = 42;
        Nd4j.dtype = DataBuffer.FLOAT;
        Random gen = new DefaultRandom(randomSeed);
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder().momentum(0.5)
                .momentumAfter(Collections.singletonMap(3, 0.9)).layerFactory(LayerFactories.getFactory(RBM.class))
                .optimizationAlgo(OptimizationAlgorithm.ITERATION_GRADIENT_DESCENT)
                .iterations(1).weightInit(WeightInit.SIZE)
                .applySparsity(true).sparsity(0.1)
                .iterationListener(new ScoreIterationListener(10))
                .lossFunction(LossFunctions.LossFunction.RECONSTRUCTION_CROSSENTROPY).rng(gen)
                .learningRate(1e-1f).nIn(784).nOut(10).list(4)
                .hiddenLayerSizes(new int[]{500, 250, 200})
                .override(new ClassifierOverride(3))
                .build();


        MultiLayerNetwork d = new MultiLayerNetwork(conf);
        DataSetIterator iter = new MultipleEpochsIterator(3,new MnistDataSetIterator(10,60000));
        while(iter.hasNext()) {
            DataSet next = iter.next();
            d.fit(next);
            next.getFeatureMatrix().data().destroy();
            next.getLabels().data().destroy();
        }


        iter.reset();

        Evaluation eval = new Evaluation();

        while(iter.hasNext()) {

            DataSet d2 = iter.next();
            INDArray predict2 = d.output(d2.getFeatureMatrix());

            eval.eval(d2.getLabels(), predict2);

        }

        log.info(eval.stats());
    }

}
```

Error log (generated by Java 1.8 configuration)
---
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00000001247f6edb, pid=6075, tid=4867
#
# JRE version: Java(TM) SE Runtime Environment (8.0_31-b13) (build 1.8.0_31-b13)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.31-b07 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libcuda_310.41.15_mercury.dylib+0xe6edb]  memobjFreeInternal+0xbb
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

---------------  T H R E A D  ---------------

Current thread (0x00007f846a801800):  JavaThread "main" [_thread_in_native, id=4867, stack(0x000000010404a000,0x000000010414a000)]

siginfo: si_signo: 11 (SIGSEGV), si_code: 0 (unknown), si_addr: 0x0000000000000000

Registers:
RAX=0x00007f846b102420, RBX=0x00007f84723827b0, RCX=0xe000000000000000, RDX=0xe000000000000000
RSP=0x0000000104148e60, RBP=0x0000000104148e70, RSI=0x0000000000000000, RDI=0x0000000000000000
R8 =0x0000000104148dc0, R9 =0x000000010414a000, R10=0xffffffffffffffff, R11=0x0000000000000000
R12=0x00007f846a95ca00, R13=0x000000076bf8bdb8, R14=0x0000000104148eb8, R15=0x00007f8468f33718
RIP=0x00000001247f6edb, EFLAGS=0x0000000000010286, ERR=0x0000000000000000
  TRAPNO=0x000000000000000d

Top of Stack: (sp=0x0000000104148e60)
0x0000000104148e60:   00007f846b102420 00007f84723827b0
0x0000000104148e70:   0000000104148ea0 00000001247f6de2
0x0000000104148e80:   0000000124a9b290 0000000000000000
0x0000000104148e90:   00007f846a95ca10 00007f84723827b0
0x0000000104148ea0:   0000000104148ee0 000000012472e944
0x0000000104148eb0:   00007f846b454d08 00007f84723827b0
0x0000000104148ec0:   0000000200c82000 0000000000000000
0x0000000104148ed0:   00007f8468fad650 00007f846a801800
0x0000000104148ee0:   0000000104148f80 00000001247159e6
0x0000000104148ef0:   0000000104148f20 0000000000000001
0x0000000104148f00:   0000000000000000 000000076bf8bdb8
0x0000000104148f10:   0000000104148f78 00007f846b454b00
0x0000000104148f20:   0000000104148f60 00000001298997da
0x0000000104148f30:   00007f8468fad650 00007f846a801800
0x0000000104148f40:   00007f846b454c20 0000000200c82000
0x0000000104148f50:   00007f8468fad650 00007f846a801800
0x0000000104148f60:   0000000104148f80 000000012989522f
0x0000000104148f70:   000003e704149100 0000000200c82000
0x0000000104148f80:   0000000104148fa0 0000000129889cf1
0x0000000104148f90:   0000000104148fa0 0000000200c82000
0x0000000104148fa0:   0000000104149070 00000001298a5bcc
0x0000000104148fb0:   0000000121dbe793 00007f8468f10e60
0x0000000104148fc0:   0000000000000001 00007f8472300000
0x0000000104148fd0:   0000000104149100 00007f846a801800
0x0000000104148fe0:   0000000104149020 0000000121dbbe0b
0x0000000104148ff0:   0000000104149100 00007f8472374ac0
0x0000000104149000:   0000000004149020 0000000121dbc485
0x0000000104149010:   00007f846a8019e8 00000001061883e0
0x0000000104149020:   00000001041490d0 0000000000000062
0x0000000104149030:   00000001041490d0 0000000000000000
0x0000000104149040:   0000000004149080 000000011f813850
0x0000000104149050:   0000000000000000 000000076bf8bdb8 

Instructions: (pc=0x00000001247f6edb)
0x00000001247f6ebb:   00 00 00 48 c7 43 28 00 00 00 00 48 8b 45 f0 48
0x00000001247f6ecb:   85 c0 74 42 48 8b 4b 08 48 85 c9 74 06 48 8b 13
0x00000001247f6edb:   48 89 11 48 8b 0b 48 85 c9 74 08 48 8b 53 08 48
0x00000001247f6eeb:   89 51 08 48 39 98 a8 01 00 00 75 0b 48 8b 4b 08 

Register to memory mapping:

RAX=0x00007f846b102420 is an unknown value
RBX=0x00007f84723827b0 is an unknown value
RCX=0xe000000000000000 is an unknown value
RDX=0xe000000000000000 is an unknown value
RSP=0x0000000104148e60 is pointing into the stack for thread: 0x00007f846a801800
RBP=0x0000000104148e70 is pointing into the stack for thread: 0x00007f846a801800
RSI=0x0000000000000000 is an unknown value
RDI=0x0000000000000000 is an unknown value
R8 =0x0000000104148dc0 is pointing into the stack for thread: 0x00007f846a801800
R9 =0x000000010414a000 is pointing into the stack for thread: 0x00007f846a801800
R10=0xffffffffffffffff is an unknown value
R11=0x0000000000000000 is an unknown value
R12=0x00007f846a95ca00 is an unknown value
R13=0x000000076bf8bdb8 is an oop
org.nd4j.linalg.jcublas.rng.distribution.UniformDistribution 
 - klass: 'org/nd4j/linalg/jcublas/rng/distribution/UniformDistribution'
R14=0x0000000104148eb8 is pointing into the stack for thread: 0x00007f846a801800
R15=0x00007f8468f33718 is an unknown value


Stack: [0x000000010404a000,0x000000010414a000],  sp=0x0000000104148e60,  free space=1019k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libcuda_310.41.15_mercury.dylib+0xe6edb]  memobjFreeInternal+0xbb
C  [libcuda_310.41.15_mercury.dylib+0xe6de2]  memobjFreeNoSync+0x62
C  [libcuda_310.41.15_mercury.dylib+0x1e944]  cuapiMemFree_common+0x164
C  [libcuda_310.41.15_mercury.dylib+0x59e6]  cuMemFree_v2+0x156
C  [libcublas.6.5.dylib+0x3cf1]
C  [libcublas.6.5.dylib+0x1fbcc]
C  [libcublas.6.5.dylib+0x4832e]  cublasFree+0xe
C  [libJCublas-apple-x86_648517511985149472288jnilib+0x34a9]  Java_jcuda_jcublas_JCublas_cublasFreeNative+0x79
J 519  jcuda.jcublas.JCublas.cublasFreeNative(Ljcuda/Pointer;)I (0 bytes) @ 0x00000001063770bf [0x0000000106377000+0xbf]
J 537% C2 org.nd4j.linalg.factory.BaseNDArrayFactory.rand([ILorg/nd4j/linalg/api/rng/distribution/Distribution;)Lorg/nd4j/linalg/api/ndarray/INDArray; (52 bytes) @ 0x000000010638c2d4 [0x000000010638bce0+0x5f4]

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
J 519  jcuda.jcublas.JCublas.cublasFreeNative(Ljcuda/Pointer;)I (0 bytes) @ 0x000000010637704f [0x0000000106377000+0x4f]
J 537% C2 org.nd4j.linalg.factory.BaseNDArrayFactory.rand([ILorg/nd4j/linalg/api/rng/distribution/Distribution;)Lorg/nd4j/linalg/api/ndarray/INDArray; (52 bytes) @ 0x000000010638c2d4 [0x000000010638bce0+0x5f4]
j  org.nd4j.linalg.factory.Nd4j.rand([ILorg/nd4j/linalg/api/rng/distribution/Distribution;)Lorg/nd4j/linalg/api/ndarray/INDArray;+5
j  org.deeplearning4j.nn.weights.WeightInitUtil.uniformBasedOnInAndOut([III)Lorg/nd4j/linalg/api/ndarray/INDArray;+45
j  org.deeplearning4j.nn.weights.WeightInitUtil.initWeights([ILorg/deeplearning4j/nn/weights/WeightInit;Lorg/nd4j/linalg/api/rng/distribution/Distribution;)Lorg/nd4j/linalg/api/ndarray/INDArray;+211
j  org.deeplearning4j.nn.weights.WeightInitUtil.initWeights(IILorg/deeplearning4j/nn/weights/WeightInit;Lorg/nd4j/linalg/api/rng/distribution/Distribution;)Lorg/nd4j/linalg/api/ndarray/INDArray;+13
j  org.deeplearning4j.nn.params.DefaultParamInitializer.createWeightMatrix(Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;)Lorg/nd4j/linalg/api/ndarray/INDArray;+16
j  org.deeplearning4j.nn.params.DefaultParamInitializer.init(Ljava/util/Map;Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;)V+5
j  org.deeplearning4j.nn.params.PretrainParamInitializer.init(Ljava/util/Map;Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;)V+3
j  org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.getParams(Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;)Ljava/util/Map;+19
j  org.deeplearning4j.nn.layers.factory.DefaultLayerFactory.create(Lorg/deeplearning4j/nn/conf/NeuralNetConfiguration;)Lorg/deeplearning4j/nn/api/Layer;+8
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.init()V+176
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.initializeLayers(Lorg/nd4j/linalg/api/ndarray/INDArray;)V+82
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.setInput(Lorg/nd4j/linalg/api/ndarray/INDArray;)V+13
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.pretrain(Lorg/nd4j/linalg/api/ndarray/INDArray;)V+27
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(Lorg/nd4j/linalg/api/ndarray/INDArray;Lorg/nd4j/linalg/api/ndarray/INDArray;)V+12
j  org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(Lorg/nd4j/linalg/dataset/api/DataSet;)V+13
j  org.deeplearning4j.DBNExample.main([Ljava/lang/String;)V+221
v  ~StubRoutines::call_stub

---------------  P R O C E S S  ---------------

Java Threads: ( => current thread )
  0x00007f846aa62000 JavaThread "process reaper" daemon [_thread_blocked, id=26375, stack(0x00000001265e1000,0x0000000126609000)]
  0x00007f846a010800 JavaThread "Service Thread" daemon [_thread_blocked, id=20739, stack(0x00000001239ed000,0x0000000123aed000)]
  0x00007f846a836000 JavaThread "C1 CompilerThread3" daemon [_thread_blocked, id=20227, stack(0x00000001238ea000,0x00000001239ea000)]
  0x00007f846900b000 JavaThread "C2 CompilerThread2" daemon [_thread_blocked, id=19715, stack(0x00000001237e7000,0x00000001238e7000)]
  0x00007f846a835000 JavaThread "C2 CompilerThread1" daemon [_thread_blocked, id=19203, stack(0x00000001236e4000,0x00000001237e4000)]
  0x00007f846982b800 JavaThread "C2 CompilerThread0" daemon [_thread_blocked, id=18691, stack(0x00000001235e1000,0x00000001236e1000)]
  0x00007f8469811800 JavaThread "Signal Dispatcher" daemon [_thread_blocked, id=14611, stack(0x00000001234de000,0x00000001235de000)]
  0x00007f846900d000 JavaThread "Finalizer" daemon [_thread_in_Java, id=13571, stack(0x0000000121ba1000,0x0000000121ca1000)]
  0x00007f846a85a800 JavaThread "Reference Handler" daemon [_thread_blocked, id=13059, stack(0x0000000121a9e000,0x0000000121b9e000)]
=>0x00007f846a801800 JavaThread "main" [_thread_in_native, id=4867, stack(0x000000010404a000,0x000000010414a000)]

Other Threads:
  0x00007f846a858000 VMThread [stack: 0x000000012199b000,0x0000000121a9b000] [id=12547]
  0x00007f846a011800 WatcherThread [stack: 0x0000000123af0000,0x0000000123bf0000] [id=21251]

VM state:not at safepoint (normal execution)

VM Mutex/Monitor currently owned by a thread: None

Heap:
 PSYoungGen      total 76288K, used 13985K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 4% used [0x000000076ab00000,0x000000076ae29568,0x000000076eb00000)
  from space 10752K, 99% used [0x000000076eb00000,0x000000076f57f1d8,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 4056K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 2% used [0x00000006c0000000,0x00000006c03f6000,0x00000006cab00000)
 Metaspace       used 6911K, capacity 7024K, committed 7296K, reserved 1056768K
  class space    used 743K, capacity 784K, committed 896K, reserved 1048576K

Card table byte_map: [0x00000001154df000,0x0000000115ce0000] byte_map_base: 0x0000000111edf000

Marking Bits: (ParMarkBitMap*) 0x0000000105a9e410
 Begin Bits: [0x0000000116a4e000, 0x000000011aa4e000)
 End Bits:   [0x000000011aa4e000, 0x000000011ea4e000)

Polling page: 0x000000010418e000

CodeCache: size=245760Kb used=2490Kb max_used=2503Kb free=243269Kb
 bounds [0x000000010611f000, 0x000000010639f000, 0x000000011511f000]
 total_blobs=956 nmethods=546 adapters=325
 compilation: enabled

Compilation events (10 events):
Event: 78.985 Thread 0x00007f846a836000  542       1       java.lang.ref.Reference::access$100 (4 bytes)
Event: 78.985 Thread 0x00007f846a836000 nmethod 542 0x0000000106380610 code [0x0000000106380760, 0x0000000106380850]
Event: 78.988 Thread 0x00007f846982b800  543   !   4       java.lang.ref.ReferenceQueue::enqueue (119 bytes)
Event: 78.990 Thread 0x00007f846982b800 nmethod 543 0x00000001063848d0 code [0x0000000106384a40, 0x0000000106384f48]
Event: 79.022 Thread 0x00007f846a836000  544 % !   3       java.lang.ref.Reference$ReferenceHandler::run @ 0 (102 bytes)
Event: 79.023 Thread 0x00007f846a836000 nmethod 544% 0x0000000106385190 code [0x00000001063853c0, 0x0000000106386108]
Event: 79.023 Thread 0x00007f846a836000  545   !   3       java.lang.ref.Reference$ReferenceHandler::run (102 bytes)
Event: 79.023 Thread 0x00007f846a836000 nmethod 545 0x000000010638ef50 code [0x000000010638f180, 0x000000010638fea8]
Event: 79.031 Thread 0x00007f846900b000  546 % !   4       java.lang.ref.Reference$ReferenceHandler::run @ 0 (102 bytes)
Event: 79.035 Thread 0x00007f846900b000 nmethod 546% 0x0000000106393850 code [0x0000000106393a00, 0x0000000106394248]

GC Heap History (2 events):
Event: 78.938 GC heap before
{Heap before GC invocations=1 (full 0):
 PSYoungGen      total 76288K, used 65536K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 100% used [0x000000076ab00000,0x000000076eb00000,0x000000076eb00000)
  from space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
  to   space 10752K, 0% used [0x000000076eb00000,0x000000076eb00000,0x000000076f580000)
 ParOldGen       total 175104K, used 0K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 0% used [0x00000006c0000000,0x00000006c0000000,0x00000006cab00000)
 Metaspace       used 6904K, capacity 7024K, committed 7296K, reserved 1056768K
  class space    used 743K, capacity 784K, committed 896K, reserved 1048576K
Event: 78.984 GC heap after
Heap after GC invocations=1 (full 0):
 PSYoungGen      total 76288K, used 10748K [0x000000076ab00000, 0x0000000770000000, 0x00000007c0000000)
  eden space 65536K, 0% used [0x000000076ab00000,0x000000076ab00000,0x000000076eb00000)
  from space 10752K, 99% used [0x000000076eb00000,0x000000076f57f1d8,0x000000076f580000)
  to   space 10752K, 0% used [0x000000076f580000,0x000000076f580000,0x0000000770000000)
 ParOldGen       total 175104K, used 4056K [0x00000006c0000000, 0x00000006cab00000, 0x000000076ab00000)
  object space 175104K, 2% used [0x00000006c0000000,0x00000006c03f6000,0x00000006cab00000)
 Metaspace       used 6904K, capacity 7024K, committed 7296K, reserved 1056768K
  class space    used 743K, capacity 784K, committed 896K, reserved 1048576K
}

Deoptimization events (1 events):
Event: 79.034 Thread 0x00007f846a85a800 Uncommon trap: reason=unreached action=reinterpret pc=0x0000000106384df8 method=java.lang.ref.ReferenceQueue.enqueue(Ljava/lang/ref/Reference;)Z @ 61

Internal exceptions (10 events):
Event: 68.528 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bce7448) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 68.843 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bd25e78) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 68.844 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bd2aa58) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.098 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf6d1d8) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.100 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf73008) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.100 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf777f0) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.100 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf7c9e8) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.100 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf82a00) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.101 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf871f0) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]
Event: 69.102 Thread 0x00007f846a801800 Exception <a 'java/security/PrivilegedActionException'> (0x000000076bf8da10) thrown at [/Users/java_re/workspace/8-2-build-macosx-x86_64/jdk8u31/2394/hotspot/src/share/vm/prims/jvm.cpp, line 1312]

Events (10 events):
Event: 69.102 loading class org/nd4j/linalg/jcublas/buffer/CudaDoubleDataBuffer done
Event: 69.102 Executing VM operation: Deoptimize
Event: 69.103 Executing VM operation: Deoptimize done
Event: 78.456 Thread 0x00007f846a801800 DEOPT PACKING pc=0x0000000106387c2c sp=0x0000000104149080
Event: 78.456 Thread 0x00007f846a801800 DEOPT UNPACKING pc=0x0000000106166373 sp=0x0000000104148eb8 mode 0
Event: 78.938 Executing VM operation: ParallelGCFailedAllocation
Event: 78.984 Executing VM operation: ParallelGCFailedAllocation done
Event: 79.034 Thread 0x00007f846a85a800 Uncommon trap: trap_request=0xffffffad fr.pc=0x0000000106384df8
Event: 79.034 Thread 0x00007f846a85a800 DEOPT PACKING pc=0x0000000106384df8 sp=0x0000000121b9d8a0
Event: 79.034 Thread 0x00007f846a85a800 DEOPT UNPACKING pc=0x0000000106123f69 sp=0x0000000121b9d848 mode 2


Dynamic libraries:
0x0000000000d67000 	/System/Library/Frameworks/Cocoa.framework/Versions/A/Cocoa
0x0000000000d67000 	/System/Library/Frameworks/Security.framework/Versions/A/Security
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/ApplicationServices
0x0000000000d67000 	/usr/lib/libz.1.dylib
0x0000000000d67000 	/usr/lib/libSystem.B.dylib
0x0000000000d67000 	/usr/lib/libobjc.A.dylib
0x0000000000d67000 	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
0x0000000000d67000 	/System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
0x0000000000d67000 	/System/Library/Frameworks/AppKit.framework/Versions/C/AppKit
0x0000000000d67000 	/System/Library/Frameworks/CoreData.framework/Versions/A/CoreData
0x0000000000d67000 	/System/Library/PrivateFrameworks/RemoteViewServices.framework/Versions/A/RemoteViewServices
0x0000000000d67000 	/System/Library/PrivateFrameworks/UIFoundation.framework/Versions/A/UIFoundation
0x0000000000d67000 	/System/Library/Frameworks/IOSurface.framework/Versions/A/IOSurface
0x0000000000d67000 	/System/Library/Frameworks/AudioToolbox.framework/Versions/A/AudioToolbox
0x0000000000d67000 	/System/Library/Frameworks/AudioUnit.framework/Versions/A/AudioUnit
0x0000000000d67000 	/System/Library/PrivateFrameworks/DataDetectorsCore.framework/Versions/A/DataDetectorsCore
0x0000000000d67000 	/System/Library/PrivateFrameworks/DesktopServicesPriv.framework/Versions/A/DesktopServicesPriv
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/HIToolbox.framework/Versions/A/HIToolbox
0x0000000000d67000 	/System/Library/Frameworks/QuartzCore.framework/Versions/A/QuartzCore
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/SpeechRecognition.framework/Versions/A/SpeechRecognition
0x0000000000d67000 	/usr/lib/libauto.dylib
0x0000000000d67000 	/usr/lib/libicucore.A.dylib
0x0000000000d67000 	/usr/lib/libxml2.2.dylib
0x0000000000d67000 	/System/Library/PrivateFrameworks/CoreUI.framework/Versions/A/CoreUI
0x0000000000d67000 	/System/Library/Frameworks/CoreAudio.framework/Versions/A/CoreAudio
0x0000000000d67000 	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration
0x0000000000d67000 	/usr/lib/liblangid.dylib
0x0000000000d67000 	/System/Library/PrivateFrameworks/MultitouchSupport.framework/Versions/A/MultitouchSupport
0x0000000000d67000 	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit
0x0000000000d67000 	/usr/lib/libDiagnosticMessagesClient.dylib
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices
0x0000000000d67000 	/System/Library/PrivateFrameworks/PerformanceAnalysis.framework/Versions/A/PerformanceAnalysis
0x0000000000d67000 	/System/Library/PrivateFrameworks/GenerationalStorage.framework/Versions/A/GenerationalStorage
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL
0x0000000000d67000 	/System/Library/PrivateFrameworks/Sharing.framework/Versions/A/Sharing
0x0000000000d67000 	/System/Library/Frameworks/CoreGraphics.framework/Versions/A/CoreGraphics
0x0000000000d67000 	/System/Library/Frameworks/CoreText.framework/Versions/A/CoreText
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/ImageIO
0x0000000000d67000 	/usr/lib/libextension.dylib
0x0000000000d67000 	/System/Library/PrivateFrameworks/Backup.framework/Versions/A/Backup
0x0000000000d67000 	/usr/lib/libarchive.2.dylib
0x0000000000d67000 	/System/Library/Frameworks/CFNetwork.framework/Versions/A/CFNetwork
0x0000000000d67000 	/System/Library/Frameworks/SystemConfiguration.framework/Versions/A/SystemConfiguration
0x0000000000d67000 	/usr/lib/libCRFSuite.dylib
0x0000000000d67000 	/usr/lib/libc++.1.dylib
0x0000000000d67000 	/usr/lib/libc++abi.dylib
0x0000000000d67000 	/usr/lib/system/libcache.dylib
0x0000000000d67000 	/usr/lib/system/libcommonCrypto.dylib
0x0000000000d67000 	/usr/lib/system/libcompiler_rt.dylib
0x0000000000d67000 	/usr/lib/system/libcopyfile.dylib
0x0000000000d67000 	/usr/lib/system/libcorecrypto.dylib
0x0000000000d67000 	/usr/lib/system/libdispatch.dylib
0x0000000000d67000 	/usr/lib/system/libdyld.dylib
0x0000000000d67000 	/usr/lib/system/libkeymgr.dylib
0x0000000000d67000 	/usr/lib/system/liblaunch.dylib
0x0000000000d67000 	/usr/lib/system/libmacho.dylib
0x0000000000d67000 	/usr/lib/system/libquarantine.dylib
0x0000000000d67000 	/usr/lib/system/libremovefile.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_asl.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_blocks.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_c.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_configuration.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_coreservices.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_coretls.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_dnssd.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_info.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_kernel.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_m.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_malloc.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_network.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_networkextension.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_notify.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_platform.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_pthread.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_sandbox.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_secinit.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_stats.dylib
0x0000000000d67000 	/usr/lib/system/libsystem_trace.dylib
0x0000000000d67000 	/usr/lib/system/libunc.dylib
0x0000000000d67000 	/usr/lib/system/libunwind.dylib
0x0000000000d67000 	/usr/lib/system/libxpc.dylib
0x0000000000d67000 	/usr/lib/libbz2.1.0.dylib
0x0000000000d67000 	/usr/lib/liblzma.5.dylib
0x0000000000d67000 	/usr/lib/libbsm.0.dylib
0x0000000000d67000 	/usr/lib/libsqlite3.dylib
0x0000000000d67000 	/usr/lib/system/libkxld.dylib
0x0000000000d67000 	/usr/lib/libxar.1.dylib
0x0000000000d67000 	/usr/lib/libpam.2.dylib
0x0000000000d67000 	/usr/lib/libOpenScriptingUtil.dylib
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/FSEvents.framework/Versions/A/FSEvents
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/CarbonCore.framework/Versions/A/CarbonCore
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/Metadata.framework/Versions/A/Metadata
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/OSServices.framework/Versions/A/OSServices
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/SearchKit.framework/Versions/A/SearchKit
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/AE.framework/Versions/A/AE
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/LaunchServices.framework/Versions/A/LaunchServices
0x0000000000d67000 	/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/DictionaryServices.framework/Versions/A/DictionaryServices
0x0000000000d67000 	/System/Library/Frameworks/NetFS.framework/Versions/A/NetFS
0x0000000000d67000 	/System/Library/PrivateFrameworks/NetAuth.framework/Versions/A/NetAuth
0x0000000000d67000 	/System/Library/PrivateFrameworks/login.framework/Versions/A/Frameworks/loginsupport.framework/Versions/A/loginsupport
0x0000000000d67000 	/System/Library/PrivateFrameworks/TCC.framework/Versions/A/TCC
0x0000000000d67000 	/usr/lib/libmecabra.dylib
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/ATS.framework/Versions/A/ATS
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/ColorSync.framework/Versions/A/ColorSync
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/HIServices.framework/Versions/A/HIServices
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/LangAnalysis.framework/Versions/A/LangAnalysis
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/PrintCore.framework/Versions/A/PrintCore
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/QD.framework/Versions/A/QD
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/SpeechSynthesis.framework/Versions/A/SpeechSynthesis
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Accelerate
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vImage.framework/Versions/A/vImage
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/vecLib
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libvDSP.dylib
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libvMisc.dylib
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLAPACK.dylib
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
0x0000000000d67000 	/System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libLinearAlgebra.dylib
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/ATS.framework/Versions/A/Resources/libFontParser.dylib
0x0000000000d67000 	/System/Library/Frameworks/ApplicationServices.framework/Versions/A/Frameworks/ATS.framework/Versions/A/Resources/libFontRegistry.dylib
0x0000000000d67000 	/System/Library/PrivateFrameworks/AppleVPA.framework/Versions/A/AppleVPA
0x0000000000d67000 	/System/Library/PrivateFrameworks/AppleJPEG.framework/Versions/A/AppleJPEG
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libJPEG.dylib
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libTIFF.dylib
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libPng.dylib
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libGIF.dylib
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libJP2.dylib
0x0000000000d67000 	/System/Library/Frameworks/ImageIO.framework/Versions/A/Resources/libRadiance.dylib
0x0000000000d67000 	/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libGLU.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libGFXShared.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libGL.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libGLImage.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libCVMSPluginSupport.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenGL.framework/Versions/A/Libraries/libCoreVMClient.dylib
0x0000000000d67000 	/usr/lib/libcups.2.dylib
0x0000000000d67000 	/System/Library/Frameworks/Kerberos.framework/Versions/A/Kerberos
0x0000000000d67000 	/System/Library/Frameworks/GSS.framework/Versions/A/GSS
0x0000000000d67000 	/usr/lib/libresolv.9.dylib
0x0000000000d67000 	/usr/lib/libiconv.2.dylib
0x0000000000d67000 	/System/Library/PrivateFrameworks/Heimdal.framework/Versions/A/Heimdal
0x0000000000d67000 	/usr/lib/libheimdal-asn1.dylib
0x0000000000d67000 	/System/Library/Frameworks/OpenDirectory.framework/Versions/A/OpenDirectory
0x0000000000d67000 	/System/Library/PrivateFrameworks/CommonAuth.framework/Versions/A/CommonAuth
0x0000000000d67000 	/System/Library/Frameworks/OpenDirectory.framework/Versions/A/Frameworks/CFOpenDirectory.framework/Versions/A/CFOpenDirectory
0x0000000000d67000 	/System/Library/Frameworks/SecurityFoundation.framework/Versions/A/SecurityFoundation
0x0000000000d67000 	/System/Library/PrivateFrameworks/LanguageModeling.framework/Versions/A/LanguageModeling
0x0000000000d67000 	/usr/lib/libcmph.dylib
0x0000000000d67000 	/System/Library/Frameworks/ServiceManagement.framework/Versions/A/ServiceManagement
0x0000000000d67000 	/usr/lib/libxslt.1.dylib
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/Ink.framework/Versions/A/Ink
0x0000000000d67000 	/System/Library/Frameworks/QuartzCore.framework/Versions/A/Frameworks/CoreImage.framework/Versions/A/CoreImage
0x0000000000d67000 	/System/Library/PrivateFrameworks/CrashReporterSupport.framework/Versions/A/CrashReporterSupport
0x0000000000d67000 	/System/Library/Frameworks/OpenCL.framework/Versions/A/OpenCL
0x0000000000d67000 	/System/Library/PrivateFrameworks/FaceCore.framework/Versions/A/FaceCore
0x0000000000d67000 	/System/Library/PrivateFrameworks/Ubiquity.framework/Versions/A/Ubiquity
0x0000000000d67000 	/System/Library/PrivateFrameworks/IconServices.framework/Versions/A/IconServices
0x0000000000d67000 	/System/Library/PrivateFrameworks/ChunkingLibrary.framework/Versions/A/ChunkingLibrary
0x0000000000d67000 	/System/Library/PrivateFrameworks/Apple80211.framework/Versions/A/Apple80211
0x0000000000d67000 	/System/Library/Frameworks/CoreWLAN.framework/Versions/A/CoreWLAN
0x0000000000d67000 	/System/Library/Frameworks/IOBluetooth.framework/Versions/A/IOBluetooth
0x0000000000d67000 	/System/Library/PrivateFrameworks/CoreWiFi.framework/Versions/A/CoreWiFi
0x0000000000d67000 	/System/Library/Frameworks/CoreBluetooth.framework/Versions/A/CoreBluetooth
0x0000000000d67000 	/System/Library/PrivateFrameworks/DebugSymbols.framework/Versions/A/DebugSymbols
0x0000000000d67000 	/System/Library/PrivateFrameworks/CoreSymbolication.framework/Versions/A/CoreSymbolication
0x0000000000d67000 	/System/Library/PrivateFrameworks/Symbolication.framework/Versions/A/Symbolication
0x0000000000d67000 	/System/Library/PrivateFrameworks/SpeechRecognitionCore.framework/Versions/A/SpeechRecognitionCore
0x0000000105200000 	/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home/jre/lib/server/libjvm.dylib
0x0000000000d67000 	/usr/lib/libstdc++.6.dylib
0x000000010414c000 	/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home/jre/lib/libverify.dylib
0x000000010415a000 	/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home/jre/lib/libjava.dylib
0x0000000104198000 	/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home/jre/lib/libzip.dylib
0x0000000121ca3000 	/System/Library/Frameworks/JavaVM.framework/Frameworks/JavaRuntimeSupport.framework/JavaRuntimeSupport
0x0000000121cbb000 	/System/Library/Frameworks/JavaVM.framework/Versions/A/Frameworks/JavaNativeFoundation.framework/Versions/A/JavaNativeFoundation
0x00000001041f2000 	/System/Library/Frameworks/JavaVM.framework/Versions/A/JavaVM
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Carbon
0x0000000121cd0000 	/System/Library/PrivateFrameworks/JavaLaunching.framework/Versions/A/JavaLaunching
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/CommonPanels.framework/Versions/A/CommonPanels
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/Help.framework/Versions/A/Help
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/ImageCapture.framework/Versions/A/ImageCapture
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/OpenScripting.framework/Versions/A/OpenScripting
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/Print.framework/Versions/A/Print
0x0000000000d67000 	/System/Library/Frameworks/Carbon.framework/Versions/A/Frameworks/SecurityHI.framework/Versions/A/SecurityHI
0x000000012467f000 	/private/var/folders/sg/__vh2pjs379fcwm0my8q74_00000gn/T/libJCudaDriver-apple-x86_646073574125132679360jnilib
0x00000001246ac000 	/usr/local/cuda/lib/libcudart.6.5.dylib
0x00000001246f5000 	/usr/local/cuda/lib/libcuda.dylib
0x00000001246f8000 	/Library/Frameworks/CUDA.framework/Versions/A/CUDA
0x0000000124710000 	/Library/Frameworks/CUDA.framework/Versions/A/Libraries/libcuda_310.41.15_mercury.dylib
0x0000000124b15000 	/Library/Frameworks/CUDA.framework/Versions/A/Libraries/libgpgpucomp.dylib
0x0000000121d80000 	/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home/jre/lib/libnet.dylib
0x0000000121d97000 	/private/var/folders/sg/__vh2pjs379fcwm0my8q74_00000gn/T/libJCurand-apple-x86_647469927455308694746jnilib
0x000000012660b000 	/usr/local/cuda/lib/libcurand.6.5.dylib
0x0000000121da6000 	/private/var/folders/sg/__vh2pjs379fcwm0my8q74_00000gn/T/libJCublas-apple-x86_648517511985149472288jnilib
0x0000000129886000 	/usr/local/cuda/lib/libcublas.6.5.dylib
0x0000000121de4000 	/private/var/folders/sg/__vh2pjs379fcwm0my8q74_00000gn/T/libJCudaRuntime-apple-x86_642065880116861998768jnilib

VM Arguments:
java_command: org.deeplearning4j.DBNExample
java_class_path (initial): target/my-jar.jar
Launcher Type: SUN_STANDARD

Environment Variables:
JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_31.jdk/Contents/Home
CLASSPATH=:/Users/vupham/bin/jdepend-2.9.1/lib/jdepend-2.9.1.jar
PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/texbin:/usr/local/cuda/bin:/Users/vupham/lib/JCuda-All-0.6.5-bin-apple-x86_64/jar/
LD_LIBRARY_PATH=/usr/local/cuda/lib:/Users/vupham/lib/deepnet/cudamat:/Users/vupham/lib/JCuda-All-0.6.5-bin-apple-x86_64/dylib/:
SHELL=/bin/bash

Signal Handlers:
SIGSEGV: [libjvm.dylib+0x57a0e7], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_ONSTACK|SA_RESTART|SA_SIGINFO
SIGBUS: [libjvm.dylib+0x57a0e7], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGFPE: [libjvm.dylib+0x45af24], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGPIPE: [libjvm.dylib+0x45af24], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGXFSZ: [libjvm.dylib+0x45af24], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGILL: [libjvm.dylib+0x45af24], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGUSR1: SIG_DFL, sa_mask[0]=11011111011111100000000111000110, sa_flags=none
SIGUSR2: [libjvm.dylib+0x45aa42], sa_mask[0]=00000000000000000000000000000000, sa_flags=SA_RESTART|SA_SIGINFO
SIGHUP: [libjvm.dylib+0x459015], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGINT: [libjvm.dylib+0x459015], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGTERM: [libjvm.dylib+0x459015], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO
SIGQUIT: [libjvm.dylib+0x459015], sa_mask[0]=11111111011111110111111111111111, sa_flags=SA_RESTART|SA_SIGINFO


---------------  S Y S T E M  ---------------

OS:Bsduname:Darwin 14.1.0 Darwin Kernel Version 14.1.0: Thu Feb 26 19:26:47 PST 2015; root:xnu-2782.10.73~1/RELEASE_X86_64 x86_64
rlimit: STACK 8192k, CORE 0k, NPROC 709, NOFILE 10240, AS infinity
load average:2.77 2.04 1.91

CPU:total 8 (4 cores per cpu, 2 threads per core) family 6 model 70 stepping 1, cmov, cx8, fxsr, mmx, sse, sse2, sse3, ssse3, sse4.1, sse4.2, popcnt, avx, avx2, aes, clmul, erms, rtm, lzcnt, ht, tsc, tscinvbit, bmi1, bmi2

Memory: 4k page, physical 16777216k(3547332k free)

/proc/meminfo:


vm_info: Java HotSpot(TM) 64-Bit Server VM (25.31-b07) for bsd-amd64 JRE (1.8.0_31-b13), built on Dec 17 2014 20:45:36 by "java_re" with gcc 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.11.00)

time: Wed Mar 11 14:05:18 2015
elapsed time: 79 seconds (0d 0h 1m 19s)

```

**Comment by agibsonccc** [comment]
Could you try to pull in the latest and see what it does? Thanks!

**Comment by phvu** [comment]
I tried it and the bug in random initialization seems fixed. However I got another exception in `JCublasWrapper.gemm`, and the training score is NaN.
I am attaching the error log. I will check it carefully tomorrow.

```
00:01:51,927 INFO  ~ Training on layer 1 with 10 examples
00:01:52,300 INFO  ~ Score at iteration 0 is NaN
00:01:52,301 INFO  ~ Training on layer 2 with 10 examples
00:01:52,493 INFO  ~ Score at iteration 0 is NaN
00:01:52,494 INFO  ~ Training on layer 3 with 10 examples
00:01:52,614 INFO  ~ Score at iteration 0 is NaN
00:01:52,699 INFO  ~ Score at iteration 0 is -0.22427436709403992
00:01:53,019 INFO  ~ Training on layer 1 with 10 examples
00:01:53,356 INFO  ~ Score at iteration 0 is NaN
00:01:53,357 INFO  ~ Training on layer 2 with 10 examples
00:01:53,556 INFO  ~ Score at iteration 0 is NaN
00:01:53,557 INFO  ~ Training on layer 3 with 10 examples
00:01:53,675 INFO  ~ Score at iteration 0 is NaN
00:01:53,740 INFO  ~ Score at iteration 0 is -0.230746790766716
00:01:53,950 INFO  ~ Training on layer 1 with 10 examples
00:01:54,251 INFO  ~ Score at iteration 0 is NaN
00:01:54,252 INFO  ~ Training on layer 2 with 10 examples
00:01:54,434 INFO  ~ Score at iteration 0 is NaN
00:01:54,434 INFO  ~ Training on layer 3 with 10 examples
00:01:54,723 INFO  ~ Score at iteration 0 is NaN
Exception in thread "main" jcuda.CudaException: CUBLAS_STATUS_INTERNAL_ERROR
	at jcuda.jcublas.JCublas.checkResultBLAS(JCublas.java:157)
	at jcuda.jcublas.JCublas.cublasSgemm(JCublas.java:5739)
	at org.nd4j.linalg.jcublas.SimpleJCublas.gemm(SimpleJCublas.java:410)
	at org.nd4j.linalg.jcublas.JCublasWrapper.gemm(JCublasWrapper.java:212)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.mmuli(BaseNDArray.java:2268)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.mmul(BaseNDArray.java:2085)
	at org.deeplearning4j.nn.layers.BaseLayer.activate(BaseLayer.java:208)
	at org.deeplearning4j.nn.layers.BaseLayer.activate(BaseLayer.java:215)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.activationFromPrevLayer(MultiLayerNetwork.java:441)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.feedForward(MultiLayerNetwork.java:463)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.finetune(MultiLayerNetwork.java:1085)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1136)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1170)
	at org.deeplearning4j.DBNExample.main(DBNExample.java:56)
```

**Comment by agibsonccc** [comment]
This is an interesting one. I'm running in to that only on larger datasets. I have been testing it with iris to get baseline functionality.

I am still working on understanding some of the subtleties. I'm glad to see at least the errors are consistent across cards.

**Comment by phvu** [comment]
With the latest code, it seems ok now. The cost is not NaN, but I got a new exception in the destructor of `BaseCudaDataBuffer`:

```
15:20:24,849 INFO  ~ Training on layer 1 with 10 examples
15:20:25,260 INFO  ~ Score at iteration 0 is -46.842218017578126
15:20:25,261 INFO  ~ Training on layer 2 with 10 examples
15:20:25,452 INFO  ~ Score at iteration 0 is 0.11439323425292969
15:20:25,452 INFO  ~ Training on layer 3 with 10 examples
15:20:25,566 INFO  ~ Score at iteration 0 is -3.5801776885986327
15:20:25,640 INFO  ~ Score at iteration 0 is -0.020797765254974364
15:20:26,083 INFO  ~ Training on layer 1 with 10 examples
15:20:26,508 INFO  ~ Score at iteration 0 is -45.999578857421874
15:20:26,509 INFO  ~ Training on layer 2 with 10 examples
15:20:26,761 INFO  ~ Score at iteration 0 is 0.3162577152252197
15:20:26,761 INFO  ~ Training on layer 3 with 10 examples
15:20:26,900 INFO  ~ Score at iteration 0 is -4.030833053588867
15:20:26,978 INFO  ~ Score at iteration 0 is -0.02868111729621887
15:20:27,204 INFO  ~ Training on layer 1 with 10 examples
15:20:27,552 INFO  ~ Score at iteration 0 is -43.635284423828125
15:20:27,553 INFO  ~ Training on layer 2 with 10 examples
15:20:27,787 INFO  ~ Score at iteration 0 is 0.8179669380187988
15:20:27,788 INFO  ~ Training on layer 3 with 10 examples
15:20:28,108 INFO  ~ Score at iteration 0 is -3.6313629150390625
15:20:28,218 INFO  ~ Score at iteration 0 is -0.029739809036254884
Exception in thread "main" jcuda.CudaException: CUBLAS_STATUS_INTERNAL_ERROR
	at jcuda.jcublas.JCublas.checkResult(JCublas.java:139)
	at jcuda.jcublas.JCublas.cublasFree(JCublas.java:277)
	at org.nd4j.linalg.jcublas.buffer.BaseCudaDataBuffer.destroy(BaseCudaDataBuffer.java:287)
	at org.deeplearning4j.DBNExample.main(DBNExample.java:57)
```

**Comment by agibsonccc** [comment]
Yeah this whole thing is still a little strange. I'd be less concerned if it were my own kernels. I'm wondering if it's a double free somewhere?

**Comment by phvu** [comment]
Yes I saw it. But it is rather a hack, we should better find out why that exception happens. I am still checking.
Moreover with the new fix, it still gives NaN on some layers.

**Comment by agibsonccc** [comment]
Well in this case, we want to prevent double frees from occurring. The internal error comes from running gemm and other operations as well.

For NaNs I'm thinking it has something to do with the distributions. There aren't any arithmetic differences that I can see (I'm running the test suite and comparing it to jblas) I'm trying to figure out the delta between the features being used in deeplearning4j and the edge cases we have yet to test.

**Comment by agibsonccc** [comment]
It seems to converge on some datasets now. It has something to do with the normal distribution. I ran it on MNIST and with 1k examples each and occasionally I see:

I'm thinking of what the errors are. There are a few likely areas of interest:

Destructors/Out of memory
RNG/Distributions



xception in thread "main" jcuda.CudaException: CUDA_ERROR_ILLEGAL_ADDRESS
	at jcuda.driver.JCudaDriver.checkResult(JCudaDriver.java:288)
	at jcuda.driver.JCudaDriver.cuCtxSynchronize(JCudaDriver.java:1852)
	at org.nd4j.linalg.jcublas.kernel.KernelFunctions.invoke(KernelFunctions.java:131)
	at org.nd4j.linalg.jcublas.ops.executioner.JCudaExecutioner.invokeFunction(JCudaExecutioner.java:243)
	at org.nd4j.linalg.jcublas.ops.executioner.JCudaExecutioner.invoke(JCudaExecutioner.java:221)
	at org.nd4j.linalg.jcublas.ops.executioner.JCudaExecutioner.exec(JCudaExecutioner.java:61)
	at org.nd4j.linalg.jcublas.ops.executioner.JCudaExecutioner.execAndReturn(JCudaExecutioner.java:81)
	at org.nd4j.linalg.api.dimensionfunctions.DimensionFunctions$13.apply(DimensionFunctions.java:152)
	at org.nd4j.linalg.api.dimensionfunctions.DimensionFunctions$13.apply(DimensionFunctions.java:149)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.doDimensionWise(BaseNDArray.java:3171)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.norm1(BaseNDArray.java:3038)
	at org.deeplearning4j.optimize.solvers.BackTrackLineSearch.optimize(BackTrackLineSearch.java:200)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.optimize(BaseOptimizer.java:172)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:44)
	at org.deeplearning4j.nn.layers.BaseLayer.fit(BaseLayer.java:318)
	at org.deeplearning4j.models.featuredetectors.rbm.RBM.fit(RBM.java:409)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.pretrain(MultiLayerNetwork.java:211)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1135)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1170)
	at org.deeplearning4j.mnist.full.DBNExample.main(DBNExample.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)

**Comment by agibsonccc** [comment]
Ok new update. Since these internal errors are likely memory related, I went another route and decided to go with a reference queue and weak references. Destructors should now be invoked when the buffers become weak references.

**Comment by agibsonccc** [comment]
Hey Vu,

At this point I think the bugs are out of memory errors from storing data on the GPU.

It's more or less stable with the new garbage collection algorithm I put in there. I'm going to close this in the next few days since mini batch training seems to work. We could possibly open another issue on scaling this. Either way, I'm feeling pretty good about the results now.

**Comment by mihaisecasiu** [comment]
I may be having a similar problem.
I'm trying the minst.full example and after a while I get this ...
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -0.08806765079498291
o.d.n.m.MultiLayerNetwork - Training on layer 1 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -505.90107421875
o.d.n.m.MultiLayerNetwork - Training on layer 2 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -271.4803955078125
o.d.n.m.MultiLayerNetwork - Training on layer 3 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -108.87943115234376
o.d.n.m.MultiLayerNetwork - Finetune phase
o.d.o.l.ScoreIterationListener - Score at iteration 0 is NaN
o.d.n.m.MultiLayerNetwork - Training on layer 1 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -583.837841796875
o.d.n.m.MultiLayerNetwork - Training on layer 2 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -236.9141357421875
o.d.n.m.MultiLayerNetwork - Training on layer 3 with 10 examples
o.d.o.l.ScoreIterationListener - Score at iteration 0 is -135.47989501953126

See Score at iteration 0 is NaN and the scores for the other layers are negative and they keep decreasing. Is this the correct behaviour? 

It goes on like that and it seems to never finish. I left it running for at least a half an hour and then I stopped it. I use jcublas-6.5 on OSX10.10

I also tries minst.small and there the logs are totally different. There it actually finishes and I get some 
results at the end.  The iris example also finishes correctly.

**Comment by agibsonccc** [comment]
Is this from most recent master?

**Comment by agibsonccc** [comment]
Could you try turning off constrainGradientUnitNorm?

**Comment by mihaisecasiu** [comment]
Yes it's the master from https://github.com/deeplearning4j/deeplearning4j.git and https://github.com/SkymindIO/nd4j.git . The examples are from: https://github.com/deeplearning4j/dl4j-0.0.3.3-examples.git. 
How do I turn off constrainGradientUnitNorm? I couldn't find any reference to that in the code...

**Comment by agibsonccc** [comment]
Ensure that the builder doesn't contain constrainGradientToUnitNorm(true). It's off by default.

Also which example is it?

**Comment by mihaisecasiu** [comment]
If you mean the NeuralNetConfiguration.Builder() then it doesn't contain constrainGradientToUnitNorm(true). This is the org.deeplearning4j.mnist.full example. Exactly as I clonned it from the git 0.0.3.3-examples repository. I changed nothing.
